backend/middleware/auth.go

```diff
@@ -0,0 +1,156 @@
+				return
+			}
+
+			if !verifySession(w, r, redirect, claims) {
+				return
+			}
+
+			isAdmin, ok := claims["isAdmin"]
+			if !ok || !isAdmin.(bool) {
+				log.WithField("status", http.StatusUnauthorized).Warn("user is not admin")
+				if redirect {
+					http.Redirect(w, r, "/denied", http.StatusSeeOther)
+				} else {
+					http.Error(w, "unauthorized", http.StatusUnauthorized)
+				}
+				return
+			}
+
+			// Token is authenticated, pass it through
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func VerifyUser(redirect bool) func(http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			token, claims, err := jwtauth.FromContext(r.Context())
+
+			if err != nil {
+				log.WithError(err).Warn("no token found")
+				if redirect {
+					http.Redirect(w, r, "/login", http.StatusSeeOther)
+				} else {
+					http.Error(w, "no token found", http.StatusUnauthorized)
+				}
+				return
+			}
+
+			if token == nil || jwt.Validate(token) != nil || claims == nil {
+				log.WithField("status", http.StatusUnauthorized).Error("invalid jwt token")
+				if redirect {
+					http.Redirect(w, r, "/denied", http.StatusSeeOther)
+				} else {
+					http.Error(w, "invalid token", http.StatusUnauthorized)
+				}
+				return
+			}
+
+			if !verifySession(w, r, redirect, claims) {
+				return
+			}
+
+			// Token is authenticated, pass it through
+			next.ServeHTTP(w, r)
+		})
+	}
+}
+
+func VerifyNotLoggedIn(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		token, claims, err := jwtauth.FromContext(r.Context())
+
+		if err == nil && token != nil && jwt.Validate(token) == nil && claims != nil && verifySession(w, r, true, claims) {
+			if isAdmin, ok := claims["isAdmin"]; ok && isAdmin.(bool) {
+				log.WithField("status", http.StatusSeeOther).Info("admin already logged in")
+				http.Redirect(w, r, "/admin", http.StatusSeeOther)
+				return
+			} else {
+				log.WithField("status", http.StatusSeeOther).Info("user already logged in")
+				http.Redirect(w, r, "/", http.StatusSeeOther)
+				return
+			}
+		}
+
+		// Token is authenticated, pass it through
+		next.ServeHTTP(w, r)
+	})
+}
+
+func verifySession(w http.ResponseWriter, r *http.Request, redirect bool, claims map[string]interface{}) bool {
+	sessionId, ok := claims["sessionId"]
+	if !ok {
+		log.WithField("status", http.StatusUnauthorized).Error("sessionId missing in jwt")
+		auth.ClearCookie(w)
+		if redirect {
+			http.Redirect(w, r, "/denied", http.StatusSeeOther)
+		} else {
+			http.Error(w, "unauthorized", http.StatusUnauthorized)
+		}
+		return false
+	}
+	if _, ok := auth.ActiveSessions[sessionId.(string)]; !ok {
+		log.Warn("token expired")
+		auth.ClearCookie(w)
+		if redirect {
+			http.Redirect(w, r, "/login", http.StatusSeeOther)
+		} else {
+			http.Error(w, "token expired", http.StatusUnauthorized)
+		}
+		return false
+	}
+	return true
+}
```

Comment:
Is this whole file a move or did anything get changed?

Reply:
The only change is it got wrapped with a `redirect` argument so that we can choose to redirect to the `/denied`, `/login`, etc. pages on certain conditions. We want that redirect behavior for frontend pages (e.g. visiting `/admin` if not admin should redirect to `/denied`), but we don't want that behavior for API requests (we just want to return a 401 for a POST request so they can see that there was an error).